<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A fluent interface for manipulating GenomicRanges â€¢ plyranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A fluent interface for manipulating GenomicRanges">
<meta property="og:description" content="A dplyr-like interface for interacting with the common Bioconductor 
    classes Ranges and GenomicRanges. By providing a grammatical
    and consistent way of manipulating these classes their accessiblity for new
    Bioconductor users is hopefully increased.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">plyranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.23.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/an-introduction.html">Introduction to plyranges</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tidyomics/plyranges" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="plyranges-fluent-genomic-data-analysis-">plyranges: fluent genomic data analysis <img id="plyranges_logo" src="reference/figures/logo.png" align="right" width="125"><a class="anchor" aria-label="anchor" href="#plyranges-fluent-genomic-data-analysis-"></a>
</h1></div>
<!-- badges: start -->

<p><a href="https://www.bioconductor.org/packages/release/bioc/html/plyranges.html" class="external-link">plyranges</a> provides a consistent interface for importing and wrangling genomics data from a variety of sources. The package defines a grammar of genomic data transformation based on <code>dplyr</code> and the Bioconductor packages <code>IRanges</code>, <code>GenomicRanges</code>, and <code>rtracklayer</code>. It does this by providing a set of verbs for developing analysis pipelines based on <em>Ranges</em> objects that represent genomic regions:</p>
<ul>
<li>Modify genomic regions with the <code><a href="reference/tidyverse-reexports.html">mutate()</a></code> and <code><a href="reference/stretch.html">stretch()</a></code> functions.</li>
<li>Modify genomic regions while fixing the start/end/center coordinates with the <code>anchor_</code> family of functions.</li>
<li>Sort genomic ranges with <code><a href="reference/tidyverse-reexports.html">arrange()</a></code>.</li>
<li>Modify, subset, and aggregate genomic data with the <code><a href="reference/tidyverse-reexports.html">mutate()</a></code>, <code><a href="reference/tidyverse-reexports.html">filter()</a></code>, and <code><a href="reference/tidyverse-reexports.html">summarise()</a></code>functions.</li>
<li>Any of the above operations can be performed on partitions of the data with <code><a href="reference/tidyverse-reexports.html">group_by()</a></code>.</li>
<li>Find nearest neighbour genomic regions with the <code>join_nearest_</code> family of functions.</li>
<li>Find overlaps between ranges with the <code>join_overlaps_</code> family of functions.</li>
<li>Merge all overlapping and adjacent genomic regions with <code><a href="reference/ranges-reduce.html">reduce_ranges()</a></code>.</li>
<li>Merge the end points of all genomic regions with <code><a href="reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</li>
<li>Import and write common genomic data formats with the <code>read_/write_</code> family of functions.</li>
</ul>
<p>For more details on the features of plyranges, read the <a href="https://sa-lee.github.io/plyranges/articles/an-introduction.html" class="external-link">vignette</a>. For a complete case-study on using plyranges to combine ATAC-seq and RNA-seq results read the <a href="https://sa-lee.github.io/fluentGenomics" class="external-link"><em>fluentGenomics</em> workflow</a>.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/plyranges.html" class="external-link">plyranges</a> can be installed from the latest Bioconductor release:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("BiocManager")</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"plyranges"</span><span class="op">)</span></span></code></pre></div>
<p>To install the development version from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sa-lee/plyranges"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="quick-overview">Quick overview<a class="anchor" aria-label="anchor" href="#quick-overview"></a>
</h1>
<div class="section level2">
<h2 id="about-ranges">About <code>Ranges</code>
<a class="anchor" aria-label="anchor" href="#about-ranges"></a>
</h2>
<p><code>Ranges</code> objects can either represent sets of integers as <code>IRanges</code> (which have start, end and width attributes) or represent genomic intervals (which have additional attributes, sequence name, and strand) as <code>GRanges</code>. In addition, both types of <code>Ranges</code> can store information about their intervals as metadata columns (for example GC content over a genomic interval).</p>
<p><code>Ranges</code> objects follow the tidy data principle: each row of a <code>Ranges</code> object corresponds to an interval, while each column will represent a variable about that interval, and generally each object will represent a single unit of observation (like gene annotations).</p>
<p>We can construct a <code>IRanges</code> object from a <code>data.frame</code> with a <code>start</code> or <code>width</code> using the <code><a href="reference/ranges-construct.html">as_iranges()</a></code> method.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, width <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/ranges-construct.html">as_iranges</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; IRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;           start       end     width</span></span>
<span><span class="co">#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]         1         5         5</span></span>
<span><span class="co">#&gt;   [2]         2         6         5</span></span>
<span><span class="co">#&gt;   [3]         3         7         5</span></span>
<span><span class="co">#&gt;   [4]         4         8         5</span></span>
<span><span class="co">#&gt;   [5]         5         9         5</span></span>
<span><span class="co"># alternatively with end</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, end <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/ranges-construct.html">as_iranges</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; IRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;           start       end     width</span></span>
<span><span class="co">#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]         1         5         5</span></span>
<span><span class="co">#&gt;   [2]         2         6         5</span></span>
<span><span class="co">#&gt;   [3]         3         7         5</span></span>
<span><span class="co">#&gt;   [4]         4         8         5</span></span>
<span><span class="co">#&gt;   [5]         5         9         5</span></span></code></pre></div>
<p>We can also construct a <code>GRanges</code> object in a similar manner. Note that a <code>GRanges</code> object requires at least a seqnames column to be present in the data.frame (but not necessarily a strand column).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>                 start <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                 width <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/ranges-construct.html">as_granges</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1       1-5      *</span></span>
<span><span class="co">#&gt;   [2]     chr2       2-6      *</span></span>
<span><span class="co">#&gt;   [3]     chr2       3-7      *</span></span>
<span><span class="co">#&gt;   [4]     chr1       4-8      *</span></span>
<span><span class="co">#&gt;   [5]     chr2       5-9      *</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co"># strand can be specified with `+`, `*` (mising) and `-`</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">strand</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"*"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/ranges-construct.html">as_granges</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1       1-5      +</span></span>
<span><span class="co">#&gt;   [2]     chr2       2-6      +</span></span>
<span><span class="co">#&gt;   [3]     chr2       3-7      -</span></span>
<span><span class="co">#&gt;   [4]     chr1       4-8      -</span></span>
<span><span class="co">#&gt;   [5]     chr2       5-9      *</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="example-finding-gwas-hits-that-overlap-known-exons">Example: finding GWAS hits that overlap known exons<a class="anchor" aria-label="anchor" href="#example-finding-gwas-hits-that-overlap-known-exons"></a>
</h1>
<p>Letâ€™s look at a more a realistic example (taken from HelloRanges vignette).</p>
<p>Suppose we have two <em>GRanges</em> objects: one containing coordinates of known exons and another containing SNPs from a GWAS.</p>
<p>The first and last 5 exons are printed below, there are two additional columns corresponding to the exon name, and a score.</p>
<p>We could check the number of exons per chromosome using <code>group_by</code> and <code>summarise</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exons</span></span>
<span><span class="co">#&gt; GRanges object with 459752 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;            seqnames            ranges strand |</span></span>
<span><span class="co">#&gt;               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span></span>
<span><span class="co">#&gt;        [1]     chr1       11874-12227      + |</span></span>
<span><span class="co">#&gt;        [2]     chr1       12613-12721      + |</span></span>
<span><span class="co">#&gt;        [3]     chr1       13221-14409      + |</span></span>
<span><span class="co">#&gt;        [4]     chr1       14362-14829      - |</span></span>
<span><span class="co">#&gt;        [5]     chr1       14970-15038      - |</span></span>
<span><span class="co">#&gt;        ...      ...               ...    ... .</span></span>
<span><span class="co">#&gt;   [459748]     chrY 59338754-59338859      + |</span></span>
<span><span class="co">#&gt;   [459749]     chrY 59338754-59338859      + |</span></span>
<span><span class="co">#&gt;   [459750]     chrY 59340194-59340278      + |</span></span>
<span><span class="co">#&gt;   [459751]     chrY 59342487-59343488      + |</span></span>
<span><span class="co">#&gt;   [459752]     chrY 59342487-59343488      + |</span></span>
<span><span class="co">#&gt;                                          name     score</span></span>
<span><span class="co">#&gt;                                   &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;        [1]    NR_046018_exon_0_0_chr1_11874_f         0</span></span>
<span><span class="co">#&gt;        [2]    NR_046018_exon_1_0_chr1_12613_f         0</span></span>
<span><span class="co">#&gt;        [3]    NR_046018_exon_2_0_chr1_13221_f         0</span></span>
<span><span class="co">#&gt;        [4]    NR_024540_exon_0_0_chr1_14362_r         0</span></span>
<span><span class="co">#&gt;        [5]    NR_024540_exon_1_0_chr1_14970_r         0</span></span>
<span><span class="co">#&gt;        ...                                ...       ...</span></span>
<span><span class="co">#&gt;   [459748] NM_002186_exon_6_0_chrY_59338754_f         0</span></span>
<span><span class="co">#&gt;   [459749] NM_176786_exon_7_0_chrY_59338754_f         0</span></span>
<span><span class="co">#&gt;   [459750] NM_002186_exon_7_0_chrY_59340194_f         0</span></span>
<span><span class="co">#&gt;   [459751] NM_002186_exon_8_0_chrY_59342487_f         0</span></span>
<span><span class="co">#&gt;   [459752] NM_176786_exon_8_0_chrY_59342487_f         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 93 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="va">exons</span> <span class="op"><a href="reference/tidyverse-reexports.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/tidyverse-reexports.html">group_by</a></span><span class="op">(</span><span class="va">seqnames</span><span class="op">)</span> <span class="op"><a href="reference/tidyverse-reexports.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/tidyverse-reexports.html">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="reference/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 49 rows and 2 columns</span></span>
<span><span class="co">#&gt;                 seqnames         n</span></span>
<span><span class="co">#&gt;                    &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1                   chr1     43366</span></span>
<span><span class="co">#&gt; 2   chr1_gl000191_random        42</span></span>
<span><span class="co">#&gt; 3   chr1_gl000192_random        46</span></span>
<span><span class="co">#&gt; 4                  chr10     19420</span></span>
<span><span class="co">#&gt; 5                  chr11     24476</span></span>
<span><span class="co">#&gt; ...                  ...       ...</span></span>
<span><span class="co">#&gt; 45        chrUn_gl000222        20</span></span>
<span><span class="co">#&gt; 46        chrUn_gl000223        22</span></span>
<span><span class="co">#&gt; 47        chrUn_gl000228        85</span></span>
<span><span class="co">#&gt; 48                  chrX     18173</span></span>
<span><span class="co">#&gt; 49                  chrY      4128</span></span></code></pre></div>
<p>Next we create a column representing the transcript_id with <code>mutate</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exons</span> <span class="op">&lt;-</span> <span class="va">exons</span> <span class="op"><a href="reference/tidyverse-reexports.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/tidyverse-reexports.html">mutate</a></span><span class="op">(</span>tx_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_exon.*"</span>, <span class="st">""</span>, <span class="va">name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To find all GWAS SNPs that overlap exons, we use <code>join_overlap_inner</code>. This will create a new <em>GRanges</em> with the coordinates of SNPs that overlap exons, as well as metadata from both objects.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">olap</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/overlap-joins.html">join_overlap_inner</a></span><span class="op">(</span><span class="va">gwas</span>, <span class="va">exons</span><span class="op">)</span></span>
<span><span class="va">olap</span></span>
<span><span class="co">#&gt; GRanges object with 3439 ranges and 4 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames    ranges strand |      name.x</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1]     chr1   1079198      * |  rs11260603</span></span>
<span><span class="co">#&gt;      [2]     chr1   1247494      * |     rs12103</span></span>
<span><span class="co">#&gt;      [3]     chr1   1247494      * |     rs12103</span></span>
<span><span class="co">#&gt;      [4]     chr1   1247494      * |     rs12103</span></span>
<span><span class="co">#&gt;      [5]     chr1   1247494      * |     rs12103</span></span>
<span><span class="co">#&gt;      ...      ...       ...    ... .         ...</span></span>
<span><span class="co">#&gt;   [3435]     chrX 153764217      * |   rs1050828</span></span>
<span><span class="co">#&gt;   [3436]     chrX 153764217      * |   rs1050828</span></span>
<span><span class="co">#&gt;   [3437]     chrX 153764217      * |   rs1050828</span></span>
<span><span class="co">#&gt;   [3438]     chrX 153764217      * |   rs1050828</span></span>
<span><span class="co">#&gt;   [3439]     chrX 153764217      * |   rs1050828</span></span>
<span><span class="co">#&gt;                                          name.y     score        tx_id</span></span>
<span><span class="co">#&gt;                                     &lt;character&gt; &lt;numeric&gt;  &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1]      NR_038869_exon_2_0_chr1_1078119_f         0    NR_038869</span></span>
<span><span class="co">#&gt;      [2]   NM_001256456_exon_1_0_chr1_1247398_r         0 NM_001256456</span></span>
<span><span class="co">#&gt;      [3]   NM_001256460_exon_1_0_chr1_1247398_r         0 NM_001256460</span></span>
<span><span class="co">#&gt;      [4]   NM_001256462_exon_1_0_chr1_1247398_r         0 NM_001256462</span></span>
<span><span class="co">#&gt;      [5]   NM_001256463_exon_1_0_chr1_1247398_r         0 NM_001256463</span></span>
<span><span class="co">#&gt;      ...                                    ...       ...          ...</span></span>
<span><span class="co">#&gt;   [3435] NM_001042351_exon_9_0_chrX_153764152_r         0 NM_001042351</span></span>
<span><span class="co">#&gt;   [3436]    NM_000402_exon_9_0_chrX_153764152_r         0    NM_000402</span></span>
<span><span class="co">#&gt;   [3437] NM_001042351_exon_9_0_chrX_153764152_r         0 NM_001042351</span></span>
<span><span class="co">#&gt;   [3438]    NM_000402_exon_9_0_chrX_153764152_r         0    NM_000402</span></span>
<span><span class="co">#&gt;   [3439] NM_001042351_exon_9_0_chrX_153764152_r         0 NM_001042351</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 93 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>For each SNP we can count the number of times it overlaps a transcript.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">olap</span> <span class="op"><a href="reference/tidyverse-reexports.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/tidyverse-reexports.html">group_by</a></span><span class="op">(</span><span class="va">name.x</span>, <span class="va">tx_id</span><span class="op">)</span> <span class="op"><a href="reference/tidyverse-reexports.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/tidyverse-reexports.html">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="reference/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1619 rows and 3 columns</span></span>
<span><span class="co">#&gt;           name.x        tx_id         n</span></span>
<span><span class="co">#&gt;      &lt;character&gt;  &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1     rs10043775 NM_001271723         1</span></span>
<span><span class="co">#&gt; 2     rs10043775    NM_030793         1</span></span>
<span><span class="co">#&gt; 3        rs10078 NM_001242412         1</span></span>
<span><span class="co">#&gt; 4        rs10078    NM_020731         1</span></span>
<span><span class="co">#&gt; 5        rs10089    NM_001046         1</span></span>
<span><span class="co">#&gt; ...          ...          ...       ...</span></span>
<span><span class="co">#&gt; 1615   rs9906595 NM_001008777         1</span></span>
<span><span class="co">#&gt; 1616      rs9948    NM_017623         1</span></span>
<span><span class="co">#&gt; 1617      rs9948    NM_199078         1</span></span>
<span><span class="co">#&gt; 1618    rs995030    NM_000899         4</span></span>
<span><span class="co">#&gt; 1619    rs995030    NM_003994         4</span></span></code></pre></div>
<p>We can also generate 2bp splice sites on either side of the exon using <code>flank_left</code> and <code>flank_right</code>. We add a column indicating the side of flanking for illustrative purposes. The <code>interweave</code> function pairs the left and right ranges objects.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">left_ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ranges-flank.html">flank_left</a></span><span class="op">(</span><span class="va">exons</span>, <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">right_ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ranges-flank.html">flank_right</a></span><span class="op">(</span><span class="va">exons</span>, <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">all_ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ranges-interweave.html">interweave</a></span><span class="op">(</span><span class="va">left_ss</span>, <span class="va">right_ss</span>, .id <span class="op">=</span> <span class="st">"side"</span><span class="op">)</span></span>
<span><span class="va">all_ss</span></span>
<span><span class="co">#&gt; GRanges object with 919504 ranges and 4 metadata columns:</span></span>
<span><span class="co">#&gt;            seqnames            ranges strand |</span></span>
<span><span class="co">#&gt;               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span></span>
<span><span class="co">#&gt;        [1]     chr1       11872-11873      + |</span></span>
<span><span class="co">#&gt;        [2]     chr1       12228-12229      + |</span></span>
<span><span class="co">#&gt;        [3]     chr1       12611-12612      + |</span></span>
<span><span class="co">#&gt;        [4]     chr1       12722-12723      + |</span></span>
<span><span class="co">#&gt;        [5]     chr1       13219-13220      + |</span></span>
<span><span class="co">#&gt;        ...      ...               ...    ... .</span></span>
<span><span class="co">#&gt;   [919500]     chrY 59340279-59340280      + |</span></span>
<span><span class="co">#&gt;   [919501]     chrY 59342485-59342486      + |</span></span>
<span><span class="co">#&gt;   [919502]     chrY 59343489-59343490      + |</span></span>
<span><span class="co">#&gt;   [919503]     chrY 59342485-59342486      + |</span></span>
<span><span class="co">#&gt;   [919504]     chrY 59343489-59343490      + |</span></span>
<span><span class="co">#&gt;                                          name     score       tx_id        side</span></span>
<span><span class="co">#&gt;                                   &lt;character&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;        [1]    NR_046018_exon_0_0_chr1_11874_f         0   NR_046018        left</span></span>
<span><span class="co">#&gt;        [2]    NR_046018_exon_0_0_chr1_11874_f         0   NR_046018       right</span></span>
<span><span class="co">#&gt;        [3]    NR_046018_exon_1_0_chr1_12613_f         0   NR_046018        left</span></span>
<span><span class="co">#&gt;        [4]    NR_046018_exon_1_0_chr1_12613_f         0   NR_046018       right</span></span>
<span><span class="co">#&gt;        [5]    NR_046018_exon_2_0_chr1_13221_f         0   NR_046018        left</span></span>
<span><span class="co">#&gt;        ...                                ...       ...         ...         ...</span></span>
<span><span class="co">#&gt;   [919500] NM_002186_exon_7_0_chrY_59340194_f         0   NM_002186       right</span></span>
<span><span class="co">#&gt;   [919501] NM_002186_exon_8_0_chrY_59342487_f         0   NM_002186        left</span></span>
<span><span class="co">#&gt;   [919502] NM_002186_exon_8_0_chrY_59342487_f         0   NM_002186       right</span></span>
<span><span class="co">#&gt;   [919503] NM_176786_exon_8_0_chrY_59342487_f         0   NM_176786        left</span></span>
<span><span class="co">#&gt;   [919504] NM_176786_exon_8_0_chrY_59342487_f         0   NM_176786       right</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 93 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="learning-more">Learning more<a class="anchor" aria-label="anchor" href="#learning-more"></a>
</h1>
<ul>
<li><p>The <a href="https://sa-lee.github.io/fluentGenomics" class="external-link"><em>fluentGenomics</em> workflow</a> package shows you how to combine differential expression genes and differential chromatin accessibility peaks using plyranges. It extends the <a href="https://github.com/mikelove/plyrangesTximetaCaseStudy" class="external-link">case study</a> by Michael Love for using plyranges with <a href="https://bioconductor.org/packages/release/bioc/html/tximeta.html" class="external-link">tximeta</a>.</p></li>
<li><p>The <a href="https://github.com/sa-lee/plyrangesWorkshops" class="external-link">extended vignette in the plyrangesWorkshops package</a> has a detailed walk through of using plyranges for coverage analysis.</p></li>
<li><p>The <a href="https://bioconductor.github.io/BiocWorkshops/fluent-genomic-data-analysis-with-plyranges.html" class="external-link">Bioc 2018 Workshop book</a> has worked examples of using <code>plyranges</code> to analyse publicly available genomics data.</p></li>
</ul>
</div>
<div class="section level1">
<h1 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h1>
<p>If you found <code>plyranges</code> useful for your work please cite our <a href="http://dx.doi.org/10.1186/s13059-018-1597-8" class="external-link">paper</a>:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="sc">@</span>ARTICLE{Lee2019,</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>  title    <span class="ot">=</span> <span class="st">"plyranges: a grammar of genomic data transformation"</span>,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  author   <span class="ot">=</span> <span class="st">"Lee, Stuart and Cook, Dianne and Lawrence, Michael"</span>,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  journal  <span class="ot">=</span> <span class="st">"Genome Biol."</span>,</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  volume   <span class="ot">=</span>  <span class="dv">20</span>,</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  number   <span class="ot">=</span>  <span class="dv">1</span>,</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  pages    <span class="ot">=</span> <span class="st">"4"</span>,</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>  month    <span class="ot">=</span>  jan,</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>  year     <span class="ot">=</span>  <span class="dv">2019</span>,</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  url      <span class="ot">=</span> <span class="st">"http://dx.doi.org/10.1186/s13059-018-1597-8"</span>,</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  doi      <span class="ot">=</span> <span class="st">"10.1186/s13059-018-1597-8"</span>,</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  pmc      <span class="ot">=</span> <span class="st">"PMC6320618"</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h1>
<p>We welcome contributions from the R/Bioconductor community. We ask that contributors follow the <a href="CODE_OF_CONDUCT.html">code of conduct</a> and the guide outlined <a href="CONTRIBUTING.html">here</a>.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/plyranges" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/tidyomics/plyranges/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/tidyomics/plyranges" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing plyranges</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="http://stuartlee.org" class="external-link">Stuart Lee</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-1179-8436" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/lawremi" class="external-link">Michael Lawrence</a> <br><small class="roles"> Author, contributor </small>  </li>
<li>Dianne Cook <br><small class="roles"> Author, contributor </small>  </li>
<li>Michael Love <br><small class="roles"> Contributor, maintainer </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sa-lee/plyranges/actions?query=workflow%3AR-CMD-check-bioc" class="external-link"><img src="https://github.com/sa-lee/plyranges/workflows/R-CMD-check-bioc/badge.svg" alt="R-CMD-check-bioc"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/plyranges" class="external-link"><img src="http://www.bioconductor.org/shields/build/release/bioc/plyranges.svg" alt="BioC status"></a></li>
<li><a href="https://codecov.io/gh/sa-lee/plyranges?branch=master" class="external-link"><img src="https://codecov.io/gh/sa-lee/plyranges/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="http://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://github.com/lawremi" class="external-link">Michael Lawrence</a>, Dianne Cook, Michael Love.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
