---
title: "Common Workflow Solutions"
author: "Stuart Lee"
date: "27/11/2018"
output: 
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```


These are the solutions to the excercies found in
[Common Genomic Data Wrangling Tasks](common-tasks.html).

# Data

We need to read in three files from HelloRangesData

- exons.bed (RefSeq hg19 annotated exons)
- cpg.bed (RefSeq hg19 annotated CpG islands)
- hesc.chromHmm.bed (predicted functional elements from chromHMM program)

```{r read-bed}
suppressPackageStartupMessages(library(plyranges))
# our genome build
build <- genome_info("hg19")

exons <- system.file("extdata", "exons.bed", package="HelloRangesData") %>% 
  read_bed(genome_info = build)

cpg <- system.file("extdata", "cpg.bed", package="HelloRangesData") %>% 
  read_bed(genome_info = build)

hesc <- system.file("extdata", "hesc.chromHmm.bed", package="HelloRangesData") %>% 
  read_bed(genome_info = build)

```

## Solutions


### Filtering and mutating

1. Modify the above example so only CpG islands that are completely
within exons are included in the result.

```{r ex-01}
cpg %>% 
  mutate(is_contained = . %within% exons) %>% 
  filter(is_contained)

# or just 
# cpg %>% filter(. %within% exons)
```

2. Create a new column in exons called tx_id (the transcript
id is everything before _exon in name).   

```{r}
# we can use the sub function in base R
exons <- exons %>% 
  mutate(tx_id = sub("_exon.*", "", name))
```


### Modifying genomic regions

1. Create a new GRanges object from CpG islands that stretches the intervals
on either side by their width while leaving the centre of the interval fixed. 

```{r stretch}
cpg_stretch <- cpg %>% 
  anchor_centre() %>% 
  mutate(width = 2*width)

# alternative is to use `stretch()`
```


2. Create a new GRanges object from exons that has only non-exonic regions.

```{r gaps}
no_exon <- exons %>% complement_ranges()
```

3. Create a new GRanges object that represent 2bp canonical splice sites
on either side of exon.

```{r flanks}
# you can do this directly with mutate and anchors
left <- exons %>% 
  anchor_start() %>% 
  mutate(start = start - 2L, width = 2L)

right <- exons %>% 
  anchor_end() %>% 
  mutate(end = end + 2L,
         width = 2L)

sites <- bind_ranges(list(left = left, right = right), .id = "side")

# or with flank_* family of functions
identical(exons %>% flank_left(2), left)
identical(exons %>% flank_right(2), right)
```

### Summarising GRanges objects

1. How many ranges are there for each predicted state?
2. How many base pairs are represented in the genome for each predicted state? 
Which state has the maximum number of bases?

```{r summarise}
state_summary <- hesc %>% 
  group_by(name) %>% 
  summarise(
    n_ranges = n(), 
    n_bases = sum(width)
  )

state_summary %>% 
  as.data.frame() %>% 
  filter(n_bases == max(n_bases))
```

### Overlaps

1. Create a new GRanges object, that has exons that are completely
within Enhancers elements of hesc. How many exons are there?

```{r}
enhancers <- hesc %>% 
  filter(grepl("Enhancer", name))

exon_within <- join_overlap_inner_within(exons, enhancers)

exon_within
```

2. Use `join_overlap_intersect()` to filter exons if at least 50 per cent
of their bases are overlapped by enhancer elements.

```{r}
exons %>% 
  mutate(total_length = width) %>% 
  join_overlap_intersect(enhancers) %>% 
  filter(width / total_length > 0.5)
```

3. Count the number of each enchancer element type that are exonic.
There are several ways of doing this, but see if you can come up with a 
solution using `join_overlap_left()` + `disjoin_ranges()`. 

```{r left-join}
enhancers %>% 
  join_overlap_left(exons) %>% 
  group_by(name.x) %>% 
  disjoin_ranges(olap_any = all(!is.na(tx_id))) %>% 
  group_by(name.x) %>% 
  summarise(prop = sum(olap_any) / n())
```
