<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read a BAM file — read_bam • plyranges</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Read a BAM file — read_bam"><meta property="og:description" content="Read a BAM file"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plyranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.23.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/an-introduction.html">Introduction to plyranges</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tidyomics/plyranges" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read a BAM file</h1>
    <small class="dont-index">Source: <a href="https://github.com/tidyomics/plyranges/blob/HEAD/R/io-bam.R" class="external-link"><code>R/io-bam.R</code></a></small>
    <div class="hidden name"><code>io-bam-read.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Read a BAM file</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_bam</span><span class="op">(</span><span class="va">file</span>, index <span class="op">=</span> <span class="va">file</span>, paired <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>file</dt>
<dd><p>A connection or path to a BAM file</p></dd>


<dt>index</dt>
<dd><p>The path to the BAM index file</p></dd>


<dt>paired</dt>
<dd><p>Whether to treat alignments as paired end (TRUE)
or single end (FALSE). Default is FALSE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A DeferredGenomicRanges object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Reading a BAM file is deferred until an action
such as using <code><a href="tidyverse-reexports.html">summarise()</a></code> or <code><a href="tidyverse-reexports.html">mutate()</a></code> occurs. If paired is set to
TRUE, when alignments are loaded, the GRanges has two additional
columns called read_pair_id and read_pair_group corresponding
to paired reads and is grouped by the read_pair_group.</p>
<p>Certain verbs have different behaviour, after using <code>read_bam()</code>.</p>
<p>For <code><a href="tidyverse-reexports.html">select()</a></code> valid columns are the fields available in the
BAM file. Valid entries are qname (QNAME), flag (FLAG),
rname (RNAME), strand, pos (POS), qwidth (width of query),
mapq (MAPQ), cigar (CIGAR), mrnm (RNEXT), mpos (PNEXT), isize
(TLEN), seq (SEQ), and qual (QUAL). Any two character tags in the BAM file
are also valid.</p>
<p>For <code><a href="tidyverse-reexports.html">filter()</a></code> the following fields are valid, to select the FALSE option
place <code>!</code> in front of the field:</p><ul><li><p><code>is_paired</code> Select either unpaired (FALSE) or paired (TRUE) reads.</p></li>
<li><p><code>is_proper_pair</code> Select either improperly paired (FALSE) or properly
paired (TRUE) reads. This is dependent on the alignment software used.</p></li>
<li><p>`is_unmapped_query``	Select unmapped (TRUE) or mapped (FALSE) reads.</p></li>
<li><p><code>has_unmapped_mate</code> Select reads with mapped (FALSE) or unmapped (TRUE) mates.</p></li>
<li><p><code>is_minus_strand</code> 	Select reads aligned to plus (FALSE) or minus (TRUE) strand.</p></li>
<li><p><code>is_mate_minus_strand</code>	Select reads where mate is aligned to plus (FALSE) or
minus (TRUE) strand.</p></li>
<li><p><code>is_first_mate_read</code>	Select reads if they are the first mate (TRUE) or
not (FALSE).</p></li>
<li><p><code>is_second_mate_read</code> Select reads if they are the second mate (TRUE) or
not (FALSE).</p></li>
<li><p><code>is_secondary_alignment</code> Select reads if their alignment status is
secondary (TRUE) or not (FALSE). This might be relevant if there are
multimapping reads.</p></li>
<li><p><code>is_not_passing_quality_controls</code> Select reads that either pass
quality controls (FALSE) or that do not (TRUE).</p></li>
<li><p><code>is_duplicate</code>	Select reads that are unduplicated (FALSE) or
duplicated (TRUE). This may represent reads that are PCR or
optical duplicates.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>Rsamtools::<a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFile()</a></code>,<code><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">GenomicAlignments::readGAlignments()</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">pasillaBamSubset</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>   <span class="va">bamfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">untreated1_chr4</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>   <span class="co"># nothing is read until an action has been performed</span></span></span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">read_bam</span><span class="op">(</span><span class="va">bamfile</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>   <span class="co"># define a region of interest</span></span></span>
<span class="r-in"><span>   <span class="va">roi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chr4"</span>, start <span class="op">=</span> <span class="fl">5e5</span>, end <span class="op">=</span> <span class="fl">7e5</span><span class="op">)</span> <span class="op"><a href="tidyverse-reexports.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>            <span class="fu"><a href="ranges-construct.html">as_granges</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>   <span class="va">rng</span> <span class="op">&lt;-</span> <span class="fu">read_bam</span><span class="op">(</span><span class="va">bamfile</span><span class="op">)</span> <span class="op"><a href="tidyverse-reexports.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>            <span class="fu"><a href="tidyverse-reexports.html">select</a></span><span class="op">(</span><span class="va">mapq</span><span class="op">)</span> <span class="op"><a href="tidyverse-reexports.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>            <span class="fu"><a href="ranges-filter-overlaps.html">filter_by_overlaps</a></span><span class="op">(</span><span class="va">roi</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: pasillaBamSubset</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DeferredGenomicRanges object with 0 ranges and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    seqnames    ranges strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: no sequences</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://github.com/lawremi" class="external-link">Michael Lawrence</a>, <a href="https://dicook.org" class="external-link">Dianne Cook</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

