<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A fluent interface for manipulating GenomicRanges • plyranges</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><meta property="og:title" content="A fluent interface for manipulating GenomicRanges">
<meta property="og:description" content="A dplyr-like interface for interacting with the common Bioconductor 
    classes Ranges and GenomicRanges. By providing a grammatical
    and consistent way of manipulating these classes their accessiblity for new
    Bioconductor users is hopefully increased.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">plyranges</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/an-introduction.html">Introduction to plyranges</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/plyranges">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->

<div id="plyranges" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#plyranges" class="anchor"></a>plyranges</h1></div>
<p><code>plryanges</code> provides a consistent interface for importing and wrangling genomics data from a variety of sources. The package defines a grammar of genomic data manipulation based on <code>dplyr</code> and the Bioconductor packages <code>IRanges</code>, <code>GenomicRanges</code>, and <code>rtracklayer</code>. It does this by providing a set of verbs for developing analysis pipelines based on <em>Ranges</em> objects that represent genomic regions:</p>
<ul>
<li>Modify genomic regions with the <code><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> and <code><a href="reference/stretch.html">stretch()</a></code> functions.</li>
<li>Modify genomic regions while fixing the start/end/center coordinates with the <code>anchor_</code> family of functions.</li>
<li>Sort genomic ranges with <code><a href="http://dplyr.tidyverse.org/reference/arrange.html">arrange()</a></code>.</li>
<li>Modify, subset, and aggregate genomic data with the <code><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code>, <code><a href="http://dplyr.tidyverse.org/reference/filter.html">filter()</a></code>, and <code><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code>functions.</li>
<li>Any of the above operations can be performed on partitions of the data with <code><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>.</li>
<li>Find nearest neighbour genomic regions with the <code>join_nearest_</code> family of functions.</li>
<li>Find overlaps between ranges with the <code>join_overlaps_</code> family of functions.</li>
<li>Merge all overlapping and adjacent genomic regions with <code><a href="reference/ranges-reduce.html">reduce_ranges()</a></code>.</li>
<li>Merge the end points of all genomic regions with <code><a href="reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</li>
<li>Import and write common genomic data formats with the <code>read_/write_</code> family of functions.</li>
</ul>
<p>For more details on the features of plryanges, read the <a href="https://sa-lee.github.io/plyranges/articles/an-introduction.html">introduction vignette</a>.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>The package is currently available from Bioconductor.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"plyranges"</span>)</code></pre></div>
<p>To install the development version from GitHub:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"sa-lee/plyranges"</span>)</code></pre></div>
</div>
<div id="quick-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-overview" class="anchor"></a>Quick overview</h1>
<div id="about-ranges" class="section level2">
<h2 class="hasAnchor">
<a href="#about-ranges" class="anchor"></a>About <code>Ranges</code>
</h2>
<p><code>Ranges</code> objects can either represent sets of integers as <code>IRanges</code> (which have start, end and width attributes) or represent genomic intervals (which have additional attributes, sequence name, and strand) as <code>GRanges</code>. In addition, both types of <code>Ranges</code> can store information about their intervals as metadata columns (for example GC content over a genomic interval).</p>
<p><code>Ranges</code> objects follow the tidy data principle: each row of a <code>Ranges</code> object corresponds to an interval, while each column will represent a variable about that interval, and generally each object will represent a single unit of observation (like gene annotations).</p>
<p>We can construct a <code>IRanges</code> object from a <code>data.frame</code> with a <code>start</code> or <code>width</code> using the <code><a href="reference/ranges-construct.html">as_iranges()</a></code> method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(plyranges)
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">width =</span> <span class="dv">5</span>)
<span class="kw"><a href="reference/ranges-construct.html">as_iranges</a></span>(df)
<span class="co">#&gt; IRanges object with 5 ranges and 0 metadata columns:</span>
<span class="co">#&gt;           start       end     width</span>
<span class="co">#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">#&gt;   [1]         1         5         5</span>
<span class="co">#&gt;   [2]         2         6         5</span>
<span class="co">#&gt;   [3]         3         7         5</span>
<span class="co">#&gt;   [4]         4         8         5</span>
<span class="co">#&gt;   [5]         5         9         5</span>
<span class="co"># alternatively with end</span>
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">end =</span> <span class="dv">5</span><span class="op">:</span><span class="dv">9</span>)
<span class="kw"><a href="reference/ranges-construct.html">as_iranges</a></span>(df)
<span class="co">#&gt; IRanges object with 5 ranges and 0 metadata columns:</span>
<span class="co">#&gt;           start       end     width</span>
<span class="co">#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">#&gt;   [1]         1         5         5</span>
<span class="co">#&gt;   [2]         2         6         5</span>
<span class="co">#&gt;   [3]         3         7         5</span>
<span class="co">#&gt;   [4]         4         8         5</span>
<span class="co">#&gt;   [5]         5         9         5</span></code></pre></div>
<p>We can also construct a <code>GRanges</code> object in a similar manner. Note that a <code>GRanges</code> object requires at least a seqnames column to be present in the data.frame (but not necessarily a strand column).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr2"</span>),
                 <span class="dt">start =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,
                 <span class="dt">width =</span> <span class="dv">5</span>)
<span class="kw"><a href="reference/ranges-construct.html">as_granges</a></span>(df)
<span class="co">#&gt; GRanges object with 5 ranges and 0 metadata columns:</span>
<span class="co">#&gt;       seqnames    ranges strand</span>
<span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;   [1]     chr1       1-5      *</span>
<span class="co">#&gt;   [2]     chr2       2-6      *</span>
<span class="co">#&gt;   [3]     chr2       3-7      *</span>
<span class="co">#&gt;   [4]     chr1       4-8      *</span>
<span class="co">#&gt;   [5]     chr2       5-9      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span>
<span class="co"># strand can be specified with `+`, `*` (mising) and `-`</span>
df<span class="op">$</span>strand &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"*"</span>)
<span class="kw"><a href="reference/ranges-construct.html">as_granges</a></span>(df)
<span class="co">#&gt; GRanges object with 5 ranges and 0 metadata columns:</span>
<span class="co">#&gt;       seqnames    ranges strand</span>
<span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;   [1]     chr1       1-5      +</span>
<span class="co">#&gt;   [2]     chr2       2-6      +</span>
<span class="co">#&gt;   [3]     chr2       3-7      -</span>
<span class="co">#&gt;   [4]     chr1       4-8      -</span>
<span class="co">#&gt;   [5]     chr2       5-9      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
</div>
<div id="example-finding-gwas-hits-that-overlap-known-exons" class="section level1">
<h1 class="hasAnchor">
<a href="#example-finding-gwas-hits-that-overlap-known-exons" class="anchor"></a>Example: finding GWAS hits that overlap known exons</h1>
<p>Let’s look at a more a realistic example (taken from HelloRanges vignette).</p>
<p>Suppose we have two <em>GRanges</em> objects: one containing coordinates of known exons and another containing SNPs from a GWAS.</p>
<p>The first and last 5 exons are printed below, there are two additional columns corresponding to the exon name, and a score.</p>
<p>We could check the number of exons per chromosome using <code>group_by</code> and <code>summarise</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons 
<span class="co">#&gt; GRanges object with 459752 ranges and 2 metadata columns:</span>
<span class="co">#&gt;            seqnames            ranges strand |</span>
<span class="co">#&gt;               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span>
<span class="co">#&gt;        [1]     chr1       11874-12227      + |</span>
<span class="co">#&gt;        [2]     chr1       12613-12721      + |</span>
<span class="co">#&gt;        [3]     chr1       13221-14409      + |</span>
<span class="co">#&gt;        [4]     chr1       14362-14829      - |</span>
<span class="co">#&gt;        [5]     chr1       14970-15038      - |</span>
<span class="co">#&gt;        ...      ...               ...    ... .</span>
<span class="co">#&gt;   [459748]     chrY 59338754-59338859      + |</span>
<span class="co">#&gt;   [459749]     chrY 59338754-59338859      + |</span>
<span class="co">#&gt;   [459750]     chrY 59340194-59340278      + |</span>
<span class="co">#&gt;   [459751]     chrY 59342487-59343488      + |</span>
<span class="co">#&gt;   [459752]     chrY 59342487-59343488      + |</span>
<span class="co">#&gt;                                          name     score</span>
<span class="co">#&gt;                                   &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">#&gt;        [1]    NR_046018_exon_0_0_chr1_11874_f         0</span>
<span class="co">#&gt;        [2]    NR_046018_exon_1_0_chr1_12613_f         0</span>
<span class="co">#&gt;        [3]    NR_046018_exon_2_0_chr1_13221_f         0</span>
<span class="co">#&gt;        [4]    NR_024540_exon_0_0_chr1_14362_r         0</span>
<span class="co">#&gt;        [5]    NR_024540_exon_1_0_chr1_14970_r         0</span>
<span class="co">#&gt;        ...                                ...       ...</span>
<span class="co">#&gt;   [459748] NM_002186_exon_6_0_chrY_59338754_f         0</span>
<span class="co">#&gt;   [459749] NM_176786_exon_7_0_chrY_59338754_f         0</span>
<span class="co">#&gt;   [459750] NM_002186_exon_7_0_chrY_59340194_f         0</span>
<span class="co">#&gt;   [459751] NM_002186_exon_8_0_chrY_59342487_f         0</span>
<span class="co">#&gt;   [459752] NM_176786_exon_8_0_chrY_59342487_f         0</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 93 sequences from an unspecified genome; no seqlengths</span>
exons <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(seqnames) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">n =</span> <span class="kw"><a href="reference/n.html">n</a></span>())
<span class="co">#&gt; DataFrame with 49 rows and 2 columns</span>
<span class="co">#&gt;                 seqnames         n</span>
<span class="co">#&gt;                    &lt;Rle&gt; &lt;integer&gt;</span>
<span class="co">#&gt; 1                   chr1     43366</span>
<span class="co">#&gt; 2   chr1_gl000191_random        42</span>
<span class="co">#&gt; 3   chr1_gl000192_random        46</span>
<span class="co">#&gt; 4                  chr10     19420</span>
<span class="co">#&gt; 5                  chr11     24476</span>
<span class="co">#&gt; ...                  ...       ...</span>
<span class="co">#&gt; 45        chrUn_gl000222        20</span>
<span class="co">#&gt; 46        chrUn_gl000223        22</span>
<span class="co">#&gt; 47        chrUn_gl000228        85</span>
<span class="co">#&gt; 48                  chrX     18173</span>
<span class="co">#&gt; 49                  chrY      4128</span></code></pre></div>
<p>Next we create a column representing the transcript_id with <code>mutate</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">tx_id =</span> <span class="kw">sub</span>(<span class="st">"_exon.*"</span>, <span class="st">""</span>, name))</code></pre></div>
<p>To find all GWAS SNPs that overlap exons, we use <code>join_overlap_inner</code>. This will create a new <em>GRanges</em> with the coordinates of SNPs that overlap exons, as well as metadata from both objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olap &lt;-<span class="st"> </span><span class="kw"><a href="reference/overlap-joins.html">join_overlap_inner</a></span>(gwas, exons)
olap
<span class="co">#&gt; GRanges object with 3439 ranges and 4 metadata columns:</span>
<span class="co">#&gt;          seqnames    ranges strand |      name.x</span>
<span class="co">#&gt;             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">#&gt;      [1]     chr1   1079198      * |  rs11260603</span>
<span class="co">#&gt;      [2]     chr1   1247494      * |     rs12103</span>
<span class="co">#&gt;      [3]     chr1   1247494      * |     rs12103</span>
<span class="co">#&gt;      [4]     chr1   1247494      * |     rs12103</span>
<span class="co">#&gt;      [5]     chr1   1247494      * |     rs12103</span>
<span class="co">#&gt;      ...      ...       ...    ... .         ...</span>
<span class="co">#&gt;   [3435]     chrX 153764217      * |   rs1050828</span>
<span class="co">#&gt;   [3436]     chrX 153764217      * |   rs1050828</span>
<span class="co">#&gt;   [3437]     chrX 153764217      * |   rs1050828</span>
<span class="co">#&gt;   [3438]     chrX 153764217      * |   rs1050828</span>
<span class="co">#&gt;   [3439]     chrX 153764217      * |   rs1050828</span>
<span class="co">#&gt;                                          name.y     score        tx_id</span>
<span class="co">#&gt;                                     &lt;character&gt; &lt;numeric&gt;  &lt;character&gt;</span>
<span class="co">#&gt;      [1]      NR_038869_exon_2_0_chr1_1078119_f         0    NR_038869</span>
<span class="co">#&gt;      [2]   NM_001256456_exon_1_0_chr1_1247398_r         0 NM_001256456</span>
<span class="co">#&gt;      [3]   NM_001256460_exon_1_0_chr1_1247398_r         0 NM_001256460</span>
<span class="co">#&gt;      [4]   NM_001256462_exon_1_0_chr1_1247398_r         0 NM_001256462</span>
<span class="co">#&gt;      [5]   NM_001256463_exon_1_0_chr1_1247398_r         0 NM_001256463</span>
<span class="co">#&gt;      ...                                    ...       ...          ...</span>
<span class="co">#&gt;   [3435] NM_001042351_exon_9_0_chrX_153764152_r         0 NM_001042351</span>
<span class="co">#&gt;   [3436]    NM_000402_exon_9_0_chrX_153764152_r         0    NM_000402</span>
<span class="co">#&gt;   [3437] NM_001042351_exon_9_0_chrX_153764152_r         0 NM_001042351</span>
<span class="co">#&gt;   [3438]    NM_000402_exon_9_0_chrX_153764152_r         0    NM_000402</span>
<span class="co">#&gt;   [3439] NM_001042351_exon_9_0_chrX_153764152_r         0 NM_001042351</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 93 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
<p>For each SNP we can count the number of times it overlaps a transcript.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olap <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(name.x, tx_id) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">n =</span> <span class="kw"><a href="reference/n.html">n</a></span>())
<span class="co">#&gt; DataFrame with 1619 rows and 3 columns</span>
<span class="co">#&gt;           name.x       tx_id         n</span>
<span class="co">#&gt;      &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span>
<span class="co">#&gt; 1     rs17121403   NM_000028         1</span>
<span class="co">#&gt; 2       rs429358   NM_000041         3</span>
<span class="co">#&gt; 3         rs7412   NM_000041         3</span>
<span class="co">#&gt; 4      rs2234978   NM_000043         1</span>
<span class="co">#&gt; 5      rs1801516   NM_000051         1</span>
<span class="co">#&gt; ...          ...         ...       ...</span>
<span class="co">#&gt; 1615   rs1420101   NR_104167         1</span>
<span class="co">#&gt; 1616   rs1065656   NR_104318         1</span>
<span class="co">#&gt; 1617   rs2224391   NR_104417         1</span>
<span class="co">#&gt; 1618   rs2224391   NR_104418         1</span>
<span class="co">#&gt; 1619  rs41281112   NR_104592         1</span></code></pre></div>
<p>We can also generate 2bp splice sites on either side of the exon using <code>flank_left</code> and <code>flank_right</code>. We add a column indicating the side of flanking for illustrative purposes. The <code>interweave</code> function pairs the left and right ranges objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">left_ss &lt;-<span class="st"> </span><span class="kw"><a href="reference/flank-ranges.html">flank_left</a></span>(exons, 2L) 
right_ss &lt;-<span class="st"> </span><span class="kw"><a href="reference/flank-ranges.html">flank_right</a></span>(exons, 2L)
all_ss &lt;-<span class="st"> </span><span class="kw"><a href="reference/ranges-interweave.html">interweave</a></span>(left_ss, right_ss, <span class="dt">.id =</span> <span class="st">"side"</span>)
all_ss
<span class="co">#&gt; GRanges object with 919504 ranges and 4 metadata columns:</span>
<span class="co">#&gt;            seqnames            ranges strand |</span>
<span class="co">#&gt;               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span>
<span class="co">#&gt;        [1]     chr1       11872-11873      + |</span>
<span class="co">#&gt;        [2]     chr1       12228-12229      + |</span>
<span class="co">#&gt;        [3]     chr1       12611-12612      + |</span>
<span class="co">#&gt;        [4]     chr1       12722-12723      + |</span>
<span class="co">#&gt;        [5]     chr1       13219-13220      + |</span>
<span class="co">#&gt;        ...      ...               ...    ... .</span>
<span class="co">#&gt;   [919500]     chrY 59340279-59340280      + |</span>
<span class="co">#&gt;   [919501]     chrY 59342485-59342486      + |</span>
<span class="co">#&gt;   [919502]     chrY 59343489-59343490      + |</span>
<span class="co">#&gt;   [919503]     chrY 59342485-59342486      + |</span>
<span class="co">#&gt;   [919504]     chrY 59343489-59343490      + |</span>
<span class="co">#&gt;                                          name     score       tx_id</span>
<span class="co">#&gt;                                   &lt;character&gt; &lt;numeric&gt; &lt;character&gt;</span>
<span class="co">#&gt;        [1]    NR_046018_exon_0_0_chr1_11874_f         0   NR_046018</span>
<span class="co">#&gt;        [2]    NR_046018_exon_0_0_chr1_11874_f         0   NR_046018</span>
<span class="co">#&gt;        [3]    NR_046018_exon_1_0_chr1_12613_f         0   NR_046018</span>
<span class="co">#&gt;        [4]    NR_046018_exon_1_0_chr1_12613_f         0   NR_046018</span>
<span class="co">#&gt;        [5]    NR_046018_exon_2_0_chr1_13221_f         0   NR_046018</span>
<span class="co">#&gt;        ...                                ...       ...         ...</span>
<span class="co">#&gt;   [919500] NM_002186_exon_7_0_chrY_59340194_f         0   NM_002186</span>
<span class="co">#&gt;   [919501] NM_002186_exon_8_0_chrY_59342487_f         0   NM_002186</span>
<span class="co">#&gt;   [919502] NM_002186_exon_8_0_chrY_59342487_f         0   NM_002186</span>
<span class="co">#&gt;   [919503] NM_176786_exon_8_0_chrY_59342487_f         0   NM_176786</span>
<span class="co">#&gt;   [919504] NM_176786_exon_8_0_chrY_59342487_f         0   NM_176786</span>
<span class="co">#&gt;                   side</span>
<span class="co">#&gt;            &lt;character&gt;</span>
<span class="co">#&gt;        [1]        left</span>
<span class="co">#&gt;        [2]       right</span>
<span class="co">#&gt;        [3]        left</span>
<span class="co">#&gt;        [4]       right</span>
<span class="co">#&gt;        [5]        left</span>
<span class="co">#&gt;        ...         ...</span>
<span class="co">#&gt;   [919500]       right</span>
<span class="co">#&gt;   [919501]        left</span>
<span class="co">#&gt;   [919502]       right</span>
<span class="co">#&gt;   [919503]        left</span>
<span class="co">#&gt;   [919504]       right</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 93 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div id="learning-more" class="section level1">
<h1 class="hasAnchor">
<a href="#learning-more" class="anchor"></a>Learning more</h1>
<p>Read the <a href="https://sa-lee.github.io/plyranges/articles/an-introduction.html">introduction vignette</a> for an overview of <code>plyranges</code> features.</p>
</div>
<div id="about-the-design-of-plyranges-api" class="section level1">
<h1 class="hasAnchor">
<a href="#about-the-design-of-plyranges-api" class="anchor"></a>About the design of plyranges API</h1>
<p>The <code>plyranges</code> package aims to provide a <a href="https://en.wikipedia.org/wiki/Fluent_interface">fluent</a> interface for performing common genomic data analysis tasks. One goal of this package is to decrease the learning curve for Bioconductor and S4 classes (especially for new users) by providing a consistent interface to common classes <em>IRanges</em> and <em>GRanges</em> (future work will extend this to the <em>SummarizedExperiment</em> class)</p>
<p>All of the methods defined in <code>plyranges</code> are type-consistent and chainable. As a consequence, users familiar with <code>dplyr</code> should be able to read and understand <code>plyranges</code> code without difficulty.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/plyranges">https://​www.bioconductor.org/​packages/​plyranges</a>
</li>
<li>Report a bug at <br><a href="https://support.bioconductor.org/">https://​support.bioconductor.org/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="http://stuartlee.org">Stuart Lee</a> <br><small class="roles"> Author, maintainer </small>  </li>
<li>
<a href="https://github.com/lawremi">Michael Lawrence</a> <br><small class="roles"> Author, contributor </small>  </li>
<li>Dianne Cook <br><small class="roles"> Author, contributor </small>  </li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/sa-lee/plyranges"><img src="https://travis-ci.org/sa-lee/plyranges.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://codecov.io/github/sa-lee/plyranges?branch=master"><img src="https://img.shields.io/codecov/c/github/sa-lee/plyranges/master.svg" alt="Coverage Status"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://stuartlee.org">Stuart Lee</a>, <a href="https://github.com/lawremi">Michael Lawrence</a>, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
