<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started with the plyranges package • plyranges</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with the plyranges package">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plyranges</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.1.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/an-introduction.html">Introduction to plyranges</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/plyranges">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Getting started with the plyranges package</h1>
                        <h4 class="author">Stuart Lee</h4>
            
            <h4 class="date">2018-09-21</h4>
      
      
      <div class="hidden name"><code>an-introduction.Rmd</code></div>

    </div>

    
    
<div id="ranges-revisited" class="section level1">
<h1 class="hasAnchor">
<a href="#ranges-revisited" class="anchor"></a><code>Ranges</code> revisited</h1>
<p>In Bioconductor there are two classes, <code>IRanges</code> and <code>GRanges</code>, that are standard data structures for representing genomics data. Throughout this document I refer to either of these classes as <code>Ranges</code> if an operation can be performed on either class, otherwise I explicitly mention if a function is appropriate for an <code>IRanges</code> or <code>GRanges</code>.</p>
<p><code>Ranges</code> objects can either represent sets of integers as <code>IRanges</code> (which have start, end and width attributes) or represent genomic intervals (which have additional attributes, sequence name, and strand) as <code>GRanges</code>. In addition, both types of <code>Ranges</code> can store information about their intervals as metadata columns (for example GC content over a genomic interval).</p>
<p><code>Ranges</code> objects follow the tidy data principle: each row of a <code>Ranges</code> object corresponds to an interval, while each column will represent a variable about that interval, and generally each object will represent a single unit of observation (like gene annotations).</p>
<p>Consequently, <code>Ranges</code> objects provide a powerful representation for reasoning about genomic data. In this vignette, you will learn more about <code>Ranges</code> objects and how via grouping, restriction and summarisation you can perform common data tasks.</p>
</div>
<div id="constructing-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#constructing-ranges" class="anchor"></a>Constructing <code>Ranges</code>
</h1>
<p>To construct an <code>IRanges</code> we require that there are at least two columns that represent at either a starting coordinate, finishing coordinate or the width of the interval.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(plyranges))</code></pre></div>
<pre><code>## Warning: package 'IRanges' was built under R version 3.5.1</code></pre>
<pre><code>## Warning: package 'S4Vectors' was built under R version 3.5.1</code></pre>
<pre><code>## Warning: package 'GenomicRanges' was built under R version 3.5.1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">100</span>)
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">2</span><span class="op">:-</span><span class="dv">1</span>, <span class="dv">13</span><span class="op">:</span><span class="dv">15</span>), 
                 <span class="dt">width=</span><span class="kw">c</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">2</span><span class="op">:</span><span class="dv">0</span>))

<span class="co"># produces IRanges</span>
rng &lt;-<span class="st"> </span>df <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>()
rng</code></pre></div>
<pre><code>## IRanges object with 7 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         1         0
##   [2]         1         1         1
##   [3]         0         1         2
##   [4]        -1         1         3
##   [5]        13        14         2
##   [6]        14        14         1
##   [7]        15        14         0</code></pre>
<p>To construct a <code>GRanges</code> we require a column that represents that sequence name ( contig or chromosome id), and an optional column to represent the strandedness of an interval.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># seqname is required for GRanges, metadata is automatically kept</span>
grng &lt;-<span class="st"> </span>df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">transform</span>(<span class="dt">seqnames =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">strand =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">gc =</span> <span class="kw">runif</span>(<span class="dv">7</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()

grng</code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr1       2-1      + |  0.76255108229816
##   [2]     chr1         1      - | 0.669021712383255
##   [3]     chr2       0-1      + | 0.204612161964178
##   [4]     chr1      -1-1      - | 0.357524853432551
##   [5]     chr1     13-14      - | 0.359475114848465
##   [6]     chr1        14      + | 0.690290528349578
##   [7]     chr2     15-14      + | 0.535811153938994
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="arithmetic-on-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#arithmetic-on-ranges" class="anchor"></a>Arithmetic on Ranges</h1>
<p>Sometimes you want to modify a genomic interval by altering the width of the interval while leaving the start, end or midpoint of the coordinates unaltered. This is achieved with the <code>mutate</code> verb along with <code>anchor_*</code> adverbs.</p>
<p>The act of anchoring fixes either the start, end, center coordinates of the <code>Range</code> object, as shown in the figure and code below and anchors are used in combination with either <code>mutate</code> or <code>stretch</code>.</p>
<p><img src="anchors.png" width="400px" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>(<span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">8</span>)))
grng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>(<span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">8</span>), 
                          <span class="dt">seqnames =</span> <span class="st">"seq1"</span>,
                          <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"-"</span>)))
<span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(rng, <span class="dt">width =</span> <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_start</a></span>(rng), <span class="dt">width =</span> <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_end</a></span>(rng), <span class="dt">width =</span> <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4         5        10
##   [2]        -7         2        10
##   [3]        -1         8        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_center</a></span>(rng), <span class="dt">width =</span> <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -2         7        10
##   [2]        -3         6        10
##   [3]         1        10        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_3p</a></span>(grng), <span class="dt">width =</span> <span class="dv">10</span>) <span class="co"># leave negative strand fixed</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      -4-5      +
##   [2]     seq1      -7-2      *
##   [3]     seq1      3-12      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_5p</a></span>(grng), <span class="dt">width =</span> <span class="dv">10</span>) <span class="co"># leave positive strand fixed</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      1-10      +
##   [2]     seq1      2-11      *
##   [3]     seq1      -1-8      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Similarly, you can modify the width of an interval using the <code>stretch</code> verb. Without anchoring, this function will extend the interval in either direction by an integer amount. With anchoring, either the start, end or midpoint are preserved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rng2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_center</a></span>(rng), <span class="dv">10</span>)
rng2</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4        10        15
##   [2]        -3         7        11
##   [3]        -2        13        16</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_end</a></span>(rng2), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]       -14        10        25
##   [2]       -13         7        21
##   [3]       -12        13        26</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_start</a></span>(rng2), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4        20        25
##   [2]        -3        17        21
##   [3]        -2        23        26</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_3p</a></span>(grng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      -9-5      +
##   [2]     seq1      -8-2      *
##   [3]     seq1      3-18      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_5p</a></span>(grng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      1-15      +
##   [2]     seq1      2-12      *
##   [3]     seq1      -7-8      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>Ranges</code> can be shifted left or right. If strand information is available we can also shift upstream or downstream.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_left</a></span>(rng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -9        -5         5
##   [2]        -8        -8         1
##   [3]        -7        -2         6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_right</a></span>(rng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        11        15         5
##   [2]        12        12         1
##   [3]        13        18         6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_upstream</a></span>(grng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1     -9--5      +
##   [2]     seq1         2      *
##   [3]     seq1     13-18      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_downstream</a></span>(grng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1     11-15      +
##   [2]     seq1         2      *
##   [3]     seq1     -7--2      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="grouping-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#grouping-ranges" class="anchor"></a>Grouping <code>Ranges</code>
</h1>
<p><code>plyranges</code> introduces a new class of <code>Ranges</code> called <code>RangesGrouped</code>, this is a similar idea to the grouped <code>data.frame\tibble</code> in <code>dplyr</code>.</p>
<p>Grouping can act on either the core components or the metadata columns of a <code>Ranges</code> object.</p>
<p>It is most effective when combined with other verbs such as <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>, <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code>, <code><a href="../reference/tidyverse-reexports.html">filter()</a></code>, <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code> or <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grng &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">strand =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">gc =</span> <span class="kw">runif</span>(<span class="dv">7</span>),
         <span class="dt">start =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>,
         <span class="dt">width =</span> <span class="dv">10</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()

grng_by_strand &lt;-<span class="st"> </span>grng <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(strand)

grng_by_strand</code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
## Groups: strand [2]
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2      1-10      - | 0.889453538926318
##   [2]     chr2      2-11      + | 0.180407245177776
##   [3]     chr2      3-12      + | 0.629390850430354
##   [4]     chr1      4-13      - | 0.989564136601985
##   [5]     chr1      5-14      + | 0.130288870073855
##   [6]     chr2      6-15      - | 0.330660525709391
##   [7]     chr2      7-16      - | 0.865120546659455
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="restricting-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#restricting-ranges" class="anchor"></a>Restricting <code>Ranges</code>
</h1>
<p>The verb <code>filter</code> can be used to restrict rows in the <code>Ranges</code>. Note that grouping will cause the <code>filter</code> to act within each group of the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grng <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(gc <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.3</span>)</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2      2-11      + | 0.180407245177776
##   [2]     chr1      5-14      + | 0.130288870073855
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filtering by group </span>
grng_by_strand <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(gc <span class="op">==</span><span class="st"> </span><span class="kw">max</span>(gc)) </code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
## Groups: strand [2]
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2      3-12      + | 0.629390850430354
##   [2]     chr1      4-13      - | 0.989564136601985
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We also provide the convenience methods <code>filter_by_overlaps</code> and <code>filter_by_non_overlaps</code> for restricting by any overlapping <code>Ranges</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>, <span class="dv">15</span>,<span class="dv">20</span>), <span class="dt">width =</span> <span class="dv">5</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>()
ir1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">6</span>, <span class="dt">width =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">7</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>()
ir0</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5
##   [3]        15        19         5
##   [4]        20        24         5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir1</code></pre></div>
<pre><code>## IRanges object with 5 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         4         3
##   [2]         3         6         4
##   [3]         4         8         5
##   [4]         5        10         6
##   [5]         6        12         7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-filter-overlaps.html">filter_by_overlaps</a></span>(ir1)</code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-filter-overlaps.html">filter_by_non_overlaps</a></span>(ir1) </code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        15        19         5
##   [2]        20        24         5</code></pre>
</div>
<div id="summarising-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#summarising-ranges" class="anchor"></a>Summarising <code>Ranges</code>
</h1>
<p>The <code>summarise</code> function will return a <code>DataFrame</code> because the information required to return a <code>Ranges</code> object is lost. It is often most useful to use <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code> in combination with the <code><a href="../reference/tidyverse-reexports.html">group_by()</a></code> family of functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir1 &lt;-<span class="st"> </span>ir1 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">gc =</span> <span class="kw">runif</span>(<span class="kw">length</span>(.)))

ir0 <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-overlaps.html">group_by_overlaps</a></span>(ir1) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(<span class="dt">gc =</span> <span class="kw">mean</span>(gc))</code></pre></div>
<pre><code>## DataFrame with 2 rows and 2 columns
##       query                gc
##   &lt;integer&gt;         &lt;numeric&gt;
## 1         1 0.675554534769617
## 2         2 0.635795166483149</code></pre>
</div>
<div id="joins-or-another-way-at-looking-at-overlaps-between-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#joins-or-another-way-at-looking-at-overlaps-between-ranges" class="anchor"></a>Joins, or another way at looking at overlaps between <code>Ranges</code>
</h1>
<p>A join acts on two GRanges objects, a query and a subject.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">query &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"chr1"</span>,
               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>),
               <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">9</span>), 
               <span class="dt">end =</span>  <span class="kw">c</span>(<span class="dv">7</span>, <span class="dv">10</span>),
               <span class="dt">key.a =</span> letters[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()

subject &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"chr1"</span>,
               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"-"</span>, <span class="st">"+"</span>),
               <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">6</span>), 
               <span class="dt">end =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">8</span>),
               <span class="dt">key.b =</span> LETTERS[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</code></pre></div>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-12-1.png" alt="Query and Subject Ranges" width="700"><p class="caption">
Query and Subject Ranges
</p>
</div>
<p>The join operator is relational in the sense that metadata from the query and subject ranges is retained in the joined range. All join operators in the <code>plyranges</code> DSL generate a set of hits based on overlap or proximity of ranges and use those hits to merge the two datasets in different ways. There are four supported matching algorithms: <em>overlap</em>, <em>nearest</em>, <em>precede</em>, and <em>follow</em>. We can further restrict the matching by whether the query is completely <em>within</em> the subject, and adding the <em>directed</em> suffix ensures that matching ranges have the same direction (strand).</p>
<p><img src="olaps.png" width="600px"></p>
<p>The first function, <code><a href="../reference/overlap-joins.html">join_overlap_intersect()</a></code> will return a <code>Ranges</code> object where the start, end, and width coordinates correspond to the amount of any overlap between the left and right input <code>Ranges</code>. It also returns any metadatain the subject range if the subject overlaps the query.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">intersect_rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_intersect</a></span>(query, subject)
intersect_rng</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1       2-4      + |        a        A
##   [2]     chr1       6-7      + |        a        B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-15-1.png" alt="Intersect Join" width="700"><p class="caption">
Intersect Join
</p>
</div>
<p>The <code><a href="../reference/overlap-joins.html">join_overlap_inner()</a></code> function will return the <code>Ranges</code> in the query that overlap any <code>Ranges</code> in the subject. Like the <code><a href="../reference/overlap-joins.html">join_overlap_intersect()</a></code> function metadata of the subject <code>Range</code> is returned if it overlaps the query.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inner_rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(query, subject)
inner_rng</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1       1-7      + |        a        A
##   [2]     chr1       1-7      + |        a        B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-17-1.png" alt="Inner Join" width="700"><p class="caption">
Inner Join
</p>
</div>
<p>We also provide a convenience method called <code>find_overlaps</code> that computes the same result as <code><a href="../reference/overlap-joins.html">join_overlap_inner()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-overlaps.html">find_overlaps</a></span>(query, subject)</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1       1-7      + |        a        A
##   [2]     chr1       1-7      + |        a        B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>The <code><a href="../reference/overlap-joins.html">join_overlap_left()</a></code> method will perform an outer left join.</p>
<p>First any overlaps that are found will be returned similar to <code><a href="../reference/overlap-joins.html">join_overlap_inner()</a></code>. Then any non-overlapping ranges will be returned, with missing values on the metadata columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">left_rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_left</a></span>(query, subject)
left_rng</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1       1-7      + |        a        A
##   [2]     chr1       1-7      + |        a        B
##   [3]     chr1      9-10      - |        b     &lt;NA&gt;
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-20-1.png" alt="Left Join" width="700"><p class="caption">
Left Join
</p>
</div>
<p>Compared with <code><a href="../reference/ranges-filter-overlaps.html">filter_by_overlaps()</a></code> above, the overlap left join expands the <code>Ranges</code> to give information about each interval on the query <code>Ranges</code> that overlap those on the subject <code>Ranges</code> as well as the intervals on the left that do not overlap any range on the right.</p>
<div id="finding-your-neighbours" class="section level2">
<h2 class="hasAnchor">
<a href="#finding-your-neighbours" class="anchor"></a>Finding your neighbours</h2>
<p>We also provide methods for finding nearest, preceding or following <code>Ranges</code>. Conceptually this is identical to our approach for finding overlaps, except the semantics of the join are different.</p>
<p><img src="neighbours.png" width="600px"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-nearest.html">join_nearest</a></span>(ir0, ir1)</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |                gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt;
##   [1]         5         9         5 | 0.780358511023223
##   [2]        10        14         5 | 0.780358511023223
##   [3]        15        19         5 | 0.780358511023223
##   [4]        20        24         5 | 0.780358511023223</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-follow.html">join_follow</a></span>(ir0, ir1)</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |                gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt;
##   [1]         5         9         5 | 0.777584439376369
##   [2]        10        14         5 | 0.603324356488883
##   [3]        15        19         5 | 0.780358511023223
##   [4]        20        24         5 | 0.780358511023223</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-precede.html">join_precede</a></span>(ir0, ir1) <span class="co"># nothing precedes returns empty `Ranges`</span></code></pre></div>
<pre><code>## IRanges object with 0 ranges and 1 metadata column:
##        start       end     width |        gc
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-precede.html">join_precede</a></span>(ir1, ir0)</code></pre></div>
<pre><code>## IRanges object with 5 ranges and 1 metadata column:
##           start       end     width |                gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt;
##   [1]         2         4         3 | 0.777584439376369
##   [2]         3         6         4 | 0.827303449623287
##   [3]         4         8         5 | 0.603324356488883
##   [4]         5        10         6 | 0.491231821943074
##   [5]         6        12         7 | 0.780358511023223</code></pre>
</div>
<div id="example-dealing-with-multimapping" class="section level2">
<h2 class="hasAnchor">
<a href="#example-dealing-with-multimapping" class="anchor"></a>Example: dealing with multimapping</h2>
<p>This example is taken from the Bioconductor support <a href="https://support.bioconductor.org/p/100046/">site</a>.</p>
<p>We have two <code>Ranges</code> objects. The first contains single nucleotide positions corresponding to an intensity measurement such as a ChiP-seq experiment, while the other contains coordinates for two genes of interest.</p>
<p>We want to identify which positions in the <code>intensities</code> <code>Ranges</code> overlap the genes, where each row corresponds to a position that overlaps a single gene.</p>
<p>First we create the two <code>Ranges</code> objects</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">intensities &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"VI"</span>,
                          <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">3320</span><span class="op">:</span><span class="dv">3321</span>,<span class="dv">3330</span><span class="op">:</span><span class="dv">3331</span>,<span class="dv">3341</span><span class="op">:</span><span class="dv">3342</span>),
                          <span class="dt">width =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()

intensities </code></pre></div>
<pre><code>## GRanges object with 6 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]       VI      3320      *
##   [2]       VI      3321      *
##   [3]       VI      3330      *
##   [4]       VI      3331      *
##   [5]       VI      3341      *
##   [6]       VI      3342      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"VI"</span>, 
                    <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">3322</span>, <span class="dv">3030</span>),
                    <span class="dt">end =</span> <span class="kw">c</span>(<span class="dv">3846</span>, <span class="dv">3338</span>),
                    <span class="dt">gene_id=</span><span class="kw">c</span>(<span class="st">"YFL064C"</span>, <span class="st">"YFL065C"</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()
                    
genes</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |  gene_id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   [1]       VI 3322-3846      * |  YFL064C
##   [2]       VI 3030-3338      * |  YFL065C
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Now to find where the positions overlap each gene, we can perform an overlap join. This will automatically carry over the gene_id information as well as their coordinates (we can drop those by only selecting the gene_id).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(intensities, genes) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">select</a></span>(gene_id)
olap</code></pre></div>
<pre><code>## GRanges object with 8 ranges and 1 metadata column:
##       seqnames    ranges strand |  gene_id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   [1]       VI      3320      * |  YFL065C
##   [2]       VI      3321      * |  YFL065C
##   [3]       VI      3330      * |  YFL065C
##   [4]       VI      3330      * |  YFL064C
##   [5]       VI      3331      * |  YFL065C
##   [6]       VI      3331      * |  YFL064C
##   [7]       VI      3341      * |  YFL064C
##   [8]       VI      3342      * |  YFL064C
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Several positions match to both genes. We can count them using <code>summarise</code> and grouping by the <code>start</code> position:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olap <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(start) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(<span class="dt">n =</span> <span class="kw"><a href="../reference/n.html">n</a></span>())</code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##       start         n
##   &lt;integer&gt; &lt;integer&gt;
## 1      3320         1
## 2      3321         1
## 3      3330         2
## 4      3331         2
## 5      3341         1
## 6      3342         1</code></pre>
</div>
<div id="grouping-by-overlaps" class="section level2">
<h2 class="hasAnchor">
<a href="#grouping-by-overlaps" class="anchor"></a>Grouping by overlaps</h2>
<p>It’s also possible to group by overlaps. Using this approach we can count the number of overlaps that are greater than 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grp_by_olap &lt;-<span class="st"> </span>ir0 <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-overlaps.html">group_by_overlaps</a></span>(ir1)
grp_by_olap</code></pre></div>
<pre><code>## IRanges object with 6 ranges and 2 metadata columns:
## Groups: query [2]
##           start       end     width |                gc     query
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt; &lt;integer&gt;
##   [1]         5         9         5 | 0.827303449623287         1
##   [2]         5         9         5 | 0.603324356488883         1
##   [3]         5         9         5 | 0.491231821943074         1
##   [4]         5         9         5 | 0.780358511023223         1
##   [5]        10        14         5 | 0.491231821943074         2
##   [6]        10        14         5 | 0.780358511023223         2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grp_by_olap <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">n_overlaps =</span> <span class="kw"><a href="../reference/n.html">n</a></span>())</code></pre></div>
<pre><code>## IRanges object with 6 ranges and 3 metadata columns:
## Groups: query [2]
##           start       end     width |                gc     query
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt; &lt;integer&gt;
##   [1]         5         9         5 | 0.827303449623287         1
##   [2]         5         9         5 | 0.603324356488883         1
##   [3]         5         9         5 | 0.491231821943074         1
##   [4]         5         9         5 | 0.780358511023223         1
##   [5]        10        14         5 | 0.491231821943074         2
##   [6]        10        14         5 | 0.780358511023223         2
##       n_overlaps
##        &lt;integer&gt;
##   [1]          4
##   [2]          4
##   [3]          4
##   [4]          4
##   [5]          2
##   [6]          2</code></pre>
<p>Of course we can also add overlap counts via the <code><a href="../reference/ranges-count-overlaps.html">count_overlaps()</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">n_overlaps =</span> <span class="kw"><a href="../reference/ranges-count-overlaps.html">count_overlaps</a></span>(., ir1))</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width | n_overlaps
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |  &lt;integer&gt;
##   [1]         5         9         5 |          4
##   [2]        10        14         5 |          2
##   [3]        15        19         5 |          0
##   [4]        20        24         5 |          0</code></pre>
</div>
</div>
<div id="data-importoutput" class="section level1">
<h1 class="hasAnchor">
<a href="#data-importoutput" class="anchor"></a>Data Import/Output</h1>
<p>We provide convenience functions via <code>rtracklayer</code> and <code>GenomicAlignments</code> for reading/writing the following data formats to/from <code>Ranges</code> objects.</p>
<table class="table">
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>File Format</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/io-bam-read.html">read_bam()</a></code></td>
<td>BAM</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/io-bed-read.html">read_bed()</a></code>/<code><a href="../reference/io-bed-write.html">write_bed()</a></code>
</td>
<td>BED</td>
</tr>
<tr class="odd">
<td>
<code>read_bedgraph()</code>/ <code>write_bedgraph()</code>
</td>
<td>BEDGraph</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/io-bed-read.html">read_narrowpeaks()</a></code>/ <code><a href="../reference/io-bed-write.html">write_narrowpeaks()</a></code>
</td>
<td>narrowPeaks</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/io-gff-read.html">read_gff()</a></code> / <code><a href="../reference/io-gff-write.html">write_gff()</a></code>
</td>
<td>GFF(1-3) / GTF</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/io-bigwig-read.html">read_bigwig()</a></code> / <code><a href="../reference/io-bigwig-write.html">write_bigwig()</a></code>
</td>
<td>BigWig</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/io-wig-read.html">read_wig()</a></code> / <code><a href="../reference/io-wig-write.html">write_wig()</a></code>
</td>
<td>Wig</td>
</tr>
</tbody>
</table>
</div>
<div id="mapping-to-genomicrangesiranges" class="section level1">
<h1 class="hasAnchor">
<a href="#mapping-to-genomicrangesiranges" class="anchor"></a>Mapping to GenomicRanges/IRanges</h1>
<p>For users already familiar with the <code>IRanges</code> and <code>GenomicRanges</code> we provide mappings to the <code>plyranges</code> API.</p>
<div id="operations-on-range-width" class="section level2">
<h2 class="hasAnchor">
<a href="#operations-on-range-width" class="anchor"></a>Operations on range width</h2>
<p>For <code>G</code>Ranges`` objects all functions ignore any strandedness, unless the strand of the range is anchored.</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="19%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>anchor_(start/end/center/3p/5p)</code></td>
<td>Fix the <code>start/end/center/</code> coordinates or positive/negative strand of range. Can be used in combination with any of the following</td>
<td>Available in functions that have a <code>fix</code> argument.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/tidyverse-reexports.html">mutate(x, width = width + modifier)</a></code></td>
<td>Modify the width of a <code>Ranges</code>
</td>
<td><code>resize</code></td>
</tr>
<tr class="odd">
<td><code><a href="../reference/stretch.html">stretch(x, extend)</a></code></td>
<td>Extend the start and end coordinates in opposite directions by a fixed amount.</td>
<td>
<code>start(x)&lt;- start(x) - extend%/%2</code>, <code>end(x) &lt;- end(x)  + extend%/%2</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="operations-on-range-width-invariant" class="section level2">
<h2 class="hasAnchor">
<a href="#operations-on-range-width-invariant" class="anchor"></a>Operations on range width (invariant)</h2>
<table class="table">
<colgroup>
<col width="33%">
<col width="19%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>shift_[left/right/downstream/upstream](x, shift)</code></td>
<td>Shift the coordinates of the interval (left/right/downstream/upstream) by an integer amount.</td>
<td>
<code>shift_right</code> corresponds to <code>shift</code>
</td>
</tr>
<tr class="even">
<td><code>flank_[left/right/downstream/upstream](x, width)</code></td>
<td>Generates flanking regions of size width <code>left/right/downstream/upstream/</code>
</td>
<td>corresponds to <code>flank</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="set-operations-vector-wise" class="section level2">
<h2 class="hasAnchor">
<a href="#set-operations-vector-wise" class="anchor"></a>Set operations (vector wise)</h2>
<p>These are usual set-operations that act on the sets of the <code>Ranges</code> represented in x and y. By default these operations will ignore any strand information. The directed versions of these functions will take into account strand.</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="19%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>[intersect/setdiff/union/]_Ranges</code></td>
<td>Set operations between two <code>Ranges</code>, ignoring strand.</td>
<td>
<code>intersect/setdiff/union/</code> with <code>ignore.strand = FALSE</code>
</td>
</tr>
<tr class="even">
<td><code>[intersect/setdiff/union/]_anchored_Ranges</code></td>
<td>As above taking into account strandedness.</td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="set-operations-element-wise" class="section level2">
<h2 class="hasAnchor">
<a href="#set-operations-element-wise" class="anchor"></a>Set operations (element wise)</h2>
<p>We provide infix operators and the verbs between and span to the represent element wise range operations. These map to the <code>pintersect/punion/psetdiff/pgap/punion(fill.gap = FALSE)</code> functions.</p>
</div>
<div id="restrictions" class="section level2">
<h2 class="hasAnchor">
<a href="#restrictions" class="anchor"></a>Restrictions</h2>
<p>The verb <code>filter</code> corresponds to <code>subset</code>, while <code>filter_by_[overlaps/non_overlaps]</code> corresponds to <code>subsetByOverlaps</code>.</p>
</div>
<div id="aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregation" class="anchor"></a>Aggregation</h2>
<p>The <code>summarise</code> verb is most similar to the <code>aggregate</code> methods defined in <code>GenomicRanges/IRanges</code>.</p>
<p>The <code>reduce_ranges/disjoin_ranges</code> correspond to the <code>reduce/disjoin</code> methods. However, the former methods allow additional summarisation.</p>
<p>The <code><a href="../reference/compute_coverage.html">compute_coverage(x)</a></code> method corresponds to <code>[I/G]Ranges(coverage(x))</code>.</p>
</div>
<div id="overlaps" class="section level2">
<h2 class="hasAnchor">
<a href="#overlaps" class="anchor"></a>Overlaps</h2>
<p>For <code>GRanges</code> objects all functions ignore any strandedness, unless the suffix <code>directed</code> is added to the function call</p>
<table style="width:100%;" class="table">
<colgroup>
<col width="25%">
<col width="54%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> function</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/ranges-overlaps.html">find_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns a <code>Ranges</code> object with any range in <code>y</code> that overlaps <code>x</code>. Appends the metadata in <code>y</code> and its genomic intervals to the returning <code>Ranges</code>.</td>
<td>
<code>findOverlaps(x,y, maxgap, minoverlap, type = "any")</code> with expanding <code>x</code> and <code>y</code> by their hits and appending the <code>mcols</code> in <code>y</code>.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/ranges-overlaps.html">group_by_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns a GroupedRanges object grouped by the query hits.</td>
<td>Same as above with an additional column called <code>query</code> which contains the queryHits.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/ranges-count-overlaps.html">count_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns an integer vector (used with <code>mutate</code>)</td>
<td><code>countOverlaps(x, y, maxgap, minoverlap, type = "any")</code></td>
</tr>
<tr class="even">
<td><code><a href="../reference/ranges-overlaps-self.html">join_overlap_self(x, maxgap, minoverlap)</a></code></td>
<td>Returns a <code>Ranges</code> object with any range that overlaps itself.</td>
<td><code>findOverlaps(x,x, maxgap, minoverlap, type = "any")</code></td>
</tr>
<tr class="odd">
<td><code><a href="../reference/overlap-joins.html">join_overlap_inner(x, y, maxgap, minoverlap)</a></code></td>
<td>Finds the intersecting <code>Ranges</code> that overlap in <code>x</code> and <code>y</code>. Returns a <code>Ranges</code> object with the metadata from <code>x</code> and <code>y</code>.</td>
<td>
<code>findOverlapsPairs(x,y, maxgap, minoverlap, type = "any")</code> + <code>pintersect</code>.</td>
</tr>
<tr class="even">
<td><code>*_within</code></td>
<td>Adding suffix <code>within</code> will find overlaps</td>
<td>Makes <code>type = "within"</code>
</td>
</tr>
<tr class="odd">
<td><code>*_includes</code></td>
<td>inverse of within functions</td>
<td>-</td>
</tr>
<tr class="even">
<td><code>join_nearest[_left/right/up/downstream](x,y)</code></td>
<td>Finds nearest neighbour <code>Ranges</code> between <code>x</code> and <code>y</code>.</td>
<td>
<code>nearest</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
<tr class="odd">
<td><code>join_precede[_left/right/up/downstream](x,y)</code></td>
<td>Finds <code>Ranges</code> in <code>x</code> that precede <code>y</code>
</td>
<td>
<code>precedes</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
<tr class="even">
<td><code>join_follow[_left/right/up/downstream](x,y)</code></td>
<td>Finds <code>Ranges</code> in <code>x</code> that follow <code>y</code>
</td>
<td>
<code>precedes</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="appendix" class="section level1">
<h1 class="hasAnchor">
<a href="#appendix" class="anchor"></a>Appendix</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.5.0 (2018-04-23)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] ggplot2_2.2.1         plyranges_1.1.9       GenomicRanges_1.33.11
## [4] GenomeInfoDb_1.17.1   IRanges_2.15.16       S4Vectors_0.19.19    
## [7] BiocGenerics_0.27.1   BiocStyle_2.9.3      
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.18                lattice_0.20-35            
##  [3] tidyr_0.8.1                 Rsamtools_1.33.3           
##  [5] Biostrings_2.49.0           assertthat_0.2.0           
##  [7] rprojroot_1.3-2             digest_0.6.15              
##  [9] plyr_1.8.4                  R6_2.2.2                   
## [11] backports_1.1.2             evaluate_0.10.1            
## [13] highr_0.7                   pillar_1.2.3               
## [15] zlibbioc_1.27.0             rlang_0.2.1                
## [17] lazyeval_0.2.1              rstudioapi_0.7             
## [19] Matrix_1.2-14               rmarkdown_1.10             
## [21] pkgdown_1.1.0               labeling_0.3               
## [23] desc_1.2.0                  BiocParallel_1.15.8        
## [25] stringr_1.3.1               munsell_0.5.0              
## [27] RCurl_1.95-4.10             DelayedArray_0.7.21        
## [29] compiler_3.5.0              rtracklayer_1.41.3         
## [31] xfun_0.2                    pkgconfig_2.0.1            
## [33] htmltools_0.3.6             tidyselect_0.2.4           
## [35] SummarizedExperiment_1.11.6 tibble_1.4.2               
## [37] GenomeInfoDbData_1.1.0      roxygen2_6.1.0             
## [39] bookdown_0.7                matrixStats_0.54.0         
## [41] XML_3.98-1.11               crayon_1.3.4               
## [43] dplyr_0.7.6                 GenomicAlignments_1.17.3   
## [45] MASS_7.3-50                 bitops_1.0-6               
## [47] commonmark_1.5              grid_3.5.0                 
## [49] gtable_0.2.0                magrittr_1.5               
## [51] scales_0.5.0                stringi_1.2.3              
## [53] reshape2_1.4.3              XVector_0.21.3             
## [55] fs_1.2.3                    bindrcpp_0.2.2             
## [57] xml2_1.2.0                  tools_3.5.0                
## [59] Biobase_2.41.2              glue_1.3.0                 
## [61] purrr_0.2.5                 yaml_2.1.19                
## [63] colorspace_1.3-2            memoise_1.1.0              
## [65] knitr_1.20                  bindr_0.1.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#ranges-revisited"><code>Ranges</code> revisited</a></li>
      <li><a href="#constructing-ranges">Constructing <code>Ranges</code></a></li>
      <li><a href="#arithmetic-on-ranges">Arithmetic on Ranges</a></li>
      <li><a href="#grouping-ranges">Grouping <code>Ranges</code></a></li>
      <li><a href="#restricting-ranges">Restricting <code>Ranges</code></a></li>
      <li><a href="#summarising-ranges">Summarising <code>Ranges</code></a></li>
      <li>
<a href="#joins-or-another-way-at-looking-at-overlaps-between-ranges">Joins, or another way at looking at overlaps between <code>Ranges</code></a><ul class="nav nav-pills nav-stacked">
<li><a href="#finding-your-neighbours">Finding your neighbours</a></li>
      <li><a href="#example-dealing-with-multimapping">Example: dealing with multimapping</a></li>
      <li><a href="#grouping-by-overlaps">Grouping by overlaps</a></li>
      </ul>
</li>
      <li><a href="#data-importoutput">Data Import/Output</a></li>
      <li>
<a href="#mapping-to-genomicrangesiranges">Mapping to GenomicRanges/IRanges</a><ul class="nav nav-pills nav-stacked">
<li><a href="#operations-on-range-width">Operations on range width</a></li>
      <li><a href="#operations-on-range-width-invariant">Operations on range width (invariant)</a></li>
      <li><a href="#set-operations-vector-wise">Set operations (vector wise)</a></li>
      <li><a href="#set-operations-element-wise">Set operations (element wise)</a></li>
      <li><a href="#restrictions">Restrictions</a></li>
      <li><a href="#aggregation">Aggregation</a></li>
      <li><a href="#overlaps">Overlaps</a></li>
      </ul>
</li>
      <li><a href="#appendix">Appendix</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://stuartlee.org">Stuart Lee</a>, <a href="https://github.com/lawremi">Michael Lawrence</a>, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
